### ISSUE TITLE
[ENH] Extend v2 BaseModel with AdamW optimizer and additional LR schedulers

### ISSUE THREAD

**Issue by StrikerEureka34** [jdb78/pytorch-forecasting#2132]
**Is your feature request related to a problem? Please describe.**
While looking into the v2 BaseModel for foundation model work (#1959, #2051), I noticed the optimizer and scheduler support is pretty limited compared to v1.

Right now, [_get_optimizer()](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model_v2.py#L304-L310) only handles "adam" and "sgd" as strings. 
v1 already supports "adamw", "ranger", generic torch.optim class name lookup, and even callable optimizers ([v1 code here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model.py#L1401-L1484)).

Same story for [_get_scheduler()](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model_v2.py#L335-L344) only ReduceLROnPlateau and StepLR are supported. 
For fine-tuning pretrained models or even just modern training loops, you'd typically want cosine annealing or warm restarts, and v2 doesn't have those yet.

**Describe the solution you'd like**
Bring  _get_optimizer() and _get_scheduler() closer to v1 parity + add a couple of schedulers that are commonly needed:

- "adamw" in  _get_optimizer(), basically a backport from v1.
- Generic torch.optim class name lookup (v1 does hasattr(torch.optim, name) which is nice)
- Callable optimizer support (pass a class or function instead of a string)
- "cosine_annealing" → CosineAnnealingLR
- "cosine_annealing_warm_restarts" → CosineAnnealingWarmRestarts
- Tests for the new paths

All changes scoped to  _base_model_v2.py, just the two methods.

**Describe alternatives you've considered**
One option would be to handle this at the FoundationModelBase level (i.e., only add these for foundation models in a future subclass). But since v1 already supports these and they're useful for any v2 model, it makes more sense to put them in the shared BaseModel so all v2 models benefit.

Another option is to just support callable optimizers and let users pass whatever they want  but having string shortcuts for common ones ("adamw", "cosine_annealing") is much better for the typical user experience and keeps things consistent with v1.

**Additional context**
This came up while working on the foundation model integration (#1959). Models like Chronos, TTM, and Moirai all use AdamW with cosine/warmup schedules for fine-tuning, and the current v2 base doesn't support any of that. It's also tracked as part of the v2 work items (#1974).

Following the guidance of @agobbifbk in #2051 to bring HF Trainer capabilities into the v2 Lightning module, this brings the optimizer/scheduler layer to parity with what foundation models expect.

I'd be happy to start working on this and submit a PR for the same.
cc @phoeenniixx @fkiraly

**Comment by phoeenniixx** [comment]
Thanks a lot @StrikerEureka34! This is a very good idea, we still lack support for many optimizers. I think @RecreationalMath would have some good ideas about it, given he worked on something similar in `sktime`? We should add the supports to `BaseModel` as you suggested. But having a big block of `if-else` statements (like we have rn) seems untidy, maybe @RecreationalMath have some insights about how to add this more cleanly?

**Comment by StrikerEureka34** [comment]
Thanks a lot for the assignment @phoeenniixx! 
And good call on the if-else blocks and I agree, the current pattern in [`_get_optimizer()`](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model_v2.py#L304-L310) and [`_get_scheduler()`](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model_v2.py#L335-L344) already feels like it'll get messy as we add more options.

I looked at what @RecreationalMath did in sktime's [`BaseDeepClassifierPyTorch`](https://github.com/sktime/sktime/blob/main/sktime/classification/deep_learning/base/_base_torch.py#L405-L458) using a dict-based registry to map string names to optimizer/scheduler classes. That way adding a new one is just a dict entry instead of another elif. I'll follow a similar pattern here.

Rough plan for the changes:

**In [_base_model_v2.py](cci:7://file:///c:/Users/nayus/Desktop/gsoc/new_kraken/pytorch-forecasting/pytorch_forecasting/models/base/_base_model_v2.py:0:0-0:0):**  

- Replace the if-else chain in [_get_optimizer()](cci:1://file:///c:/Users/nayus/Desktop/gsoc/new_kraken/pytorch-forecasting/pytorch_forecasting/models/base/_base_model_v2.py:294:4-316:13) with a class-level registry dict:

  ```python
  _OPTIMIZER_REGISTRY = {
      "adam": torch.optim.Adam,
      "adamw": torch.optim.AdamW,
      "sgd": torch.optim.SGD,
  }

- Also add a hasattr(torch.optim, name) fallback for any optimizer not in the dict (like [v1 does here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model.py#L1458-L1469)), and callable support for passing a class directly.


- Same approach for _get_scheduler():

  ```python
  _SCHEDULER_REGISTRY = {
      "reduce_lr_on_plateau": torch.optim.lr_scheduler.ReduceLROnPlateau,
      "step_lr": torch.optim.lr_scheduler.StepLR,
      "cosine_annealing": torch.optim.lr_scheduler.CosineAnnealingLR,
      "cosine_annealing_warm_restarts": torch.optim.lr_scheduler.CosineAnnealingWarmRestarts,
  }

- New test file,  "tests/test_models/test_base_model_v2.py" :
                      - Tests for all optimizer paths (string shortcuts, generic torch.optim lookup, callable)
                      - Tests for all scheduler paths
                      - Error handling for invalid names

Looking forward to any thoughts on the registry approach, especially from @RecreationalMath! If this looks good I'll start working on the PR.

**Comment by RecreationalMath** [comment]
@StrikerEureka34 Yes this is the right approach, you can go ahead with the implementation.

Happy coding!

<ins>Update</ins>:
I believe ptf supports callbacks via `lightning` library, then why do we have a custom `_get_scheduler()` method in the base class? (schedulers are also callbacks). In sktime it is done so because we haven't yet added support for callbacks which are not native to the `torch` library.

My guess, Is this because of ptf V1/V2? one has `_get_scheduler()` method and other supports via `lightning`? @phoeenniixx

**Comment by phoeenniixx** [comment]
> My guess, Is this because of ptf V1/V2? one has `_get_scheduler()` method and other supports via `lightning`?

No, rn both the versions have their own version of this methods. Actually to configure optimizers, we need to override [`_configure_optimizers`](https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.configure_optimizers) of `LightningModule`. In v1, we do it directly in the method (see [here](https://github.com/sktime/pytorch-forecasting/blob/cf46d07162ac3ffcbbb382214fbde4659f4f4d55/pytorch_forecasting/models/base/_base_model.py#L1495)), while in v2, we call `_get_scheduler()` inside `_configure_optimizers` (see [here](https://github.com/sktime/pytorch-forecasting/blob/cf46d07162ac3ffcbbb382214fbde4659f4f4d55/pytorch_forecasting/models/base/_base_model_v2.py#L282)).
Here, the talk is about `torch` optimizers and schedulers, I am not sure you can pass directly it as "callbacks" to the `lightning`.  The callbacks to `lightning` are sent through `trainer`, but these are to be set during model configuration, yes?
So, we need to configure it in a similar way Ig. This is based on my understanding...

**Comment by RecreationalMath** [comment]
@phoeenniixx ahh, that's right, `_configure_optimizers` is the standard Lightning hook for this, and LR schedulers returned from it get managed by Lightning's training loop automatically. They're not trainer callbacks in the traditional sense.

That said, Lightning does have LearningRateMonitor as an actual Trainer callback for logging LR values, which might be worth wiring up separately at some point, but that's separate from the scheduler configuration itself.

So the current design of having `_get_scheduler()` as a helper called from `_configure_optimizers` makes sense and should stay. @StrikerEureka34, the registry approach looks good to go, feel free to start the work/PR.

Happy coding!

**Comment by StrikerEureka34** [comment]
Thanks @RecreationalMath and @phoeenniixx, the discussion on scheduler-vs-callback is much appreciated, that clears things up nicely. Good to know the current `_get_scheduler()` -> `_configure_optimizers` design is the right pattern to keep.

I'll start working on the registry-based PR now. Will keep it scoped to the two methods in `_base_model_v2.py` + tests, as outlined above.
